---
title: "Analyzing cell-free DNA methylation data with cfTools"
author: 
  - name: Ran Hu
    affiliation:
      - Department of Pathology and Laboratory Medicine, David Geffen School of Medicine, University of California at Los Angeles
      - Institute for Quantitative & Computational Biosciences, University of California at Los Angeles
      - Bioinformatics Interdepartmental Graduate Program, University of California at Los Angeles
    email: huran@ucla.edu
  - name: Wenyuan Li
    affiliation:
      - Department of Pathology and Laboratory Medicine, David Geffen School of Medicine, University of California at Los Angeles
  - name: Xianghong Jasmine Zhou
    affiliation:
      - Department of Pathology and Laboratory Medicine, David Geffen School of Medicine, University of California at Los Angeles
      - Institute for Quantitative & Computational Biosciences, University of California at Los Angeles
package: cfTools
output: 
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
abstract: >
  Cell-free DNA (cfDNA) in blood has emerged as an ideal surrogate for tumor biopsy. It can be obtained noninvasively, and provides a comprehensive landscape of the heterogeneous genetic and epigenetic alterations in tumors. The `r Biocpkg("cfTools")` package provides methods for cfDNA methylation data analysis to facilitate cfDNA-based cancer studies, including (1) cancer detection: sensitively detect tumor-derived cfDNA to determine whether a sample is from a cancer or normal patient; (2) tissue deconvolution: infer the tissue type composition of a plasma cfDNA sample. Note that this vignette is modularized and exmaple datasets used in different modules may not be relevant.
vignette: >
  %\VignetteIndexEntry{cfTools-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Given the methylation sequencing data of a patient's cell-free DNA (cfDNA) sample, for each cancer marker or tissue marker, we deconvolve the tumor-derived or tissue-specific reads from all reads falling in the marker region. Our read-based deconvolution algorithm exploits the pervasiveness of DNA methylation for signal enhancement, therefore can sensitively identify a trace amount of tumor-specific or tissue-specific cfDNA in plasma. 

`r Biocpkg("cfTools")` can deconvolve different sources of cfDNA fragments (or reads) in two contexts:

1. Cancer detection [1]: separating cfDNA fragments into tumor-derived fragments and background normal fragments (2 classes).

2. Tissue deconvolution [2]: separating cfDNA fragments from different tissues (> 2 classes). 

## Cancer detection

The goal of this method is to detect tumor-derived cfDNA. Given a set of tumor-specific methylation markers, we take all cfDNA fragments reads mapped to marker regions and calculate the probability of each read derived from tumor or normal tissues based on the joint-methylation-states of CpG sites on the fragment. The methylation states of all CpG sites covered by the fragment are represented by a sequence of binary values (0 represents unmethylated CpG and 1 represents methylated CpG). The likelihood of each fragment derived from tumor or normal class can be used to estimate the tumor-derived cfDNA fraction $\theta$ ($0\leq \theta < 1$).

## Tissue deconvolution

This method aims to infer $T$-tissue-type composition of a plasma cfDNA sample, which is denoted as a composition vector $\Theta=(\theta_1,\theta_2,\cdots,\theta_T)$ satisfying $0\leq \theta_i < 1$ and $\sum_{t=1}^{T}{\theta_t}=1$ and $\theta_t$ is the cfDNA fraction of tissue $t$ in the plasma. Given a set of $T$-tissue-type methylation signatures, we map all cfDNA sequencing fragments into the genomic regions of these methylation signatures and calculate the tissue-of-origin likelihood of each cfDNA fragment for tissue $t$. The methylation states of all CpG sites at individual fragment level could also be represented by a sequence of binary values (0 represents unmethylated CpG and 1 represents methylated CpG).


# Installation

`r Biocpkg("cfTools")` is an `R` package available via the [Bioconductor](http://bioconductor.org) repository for packages. You can install the release version by using the following commands in your `R` session:

```{r "install", eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}
BiocManager::install("cfTools")
```

Alternatively, you can install the development version from [GitHub](https://github.com/) :

```{r 'install_dev', eval = FALSE}
BiocManager::install("jasminezhoulab/cfTools")
```

# Input data preparation

Example input data files are included within the package:
```{r demo}
library(cfTools)
demo.dir <- system.file("extdata", package="cfTools")
```

## Collapse paired-end sequencing reads to fragment-level

Function `CollapsePEReads()` generates fragment-level information for paired-end sequencing reads. The main input file is a standard (sorted) BED file (e.g. output of *bedtools bamtobed*) of paired-end sequencing reads containing columns of chromosome name, chromosome start, chromosome end, sequence ID, score, and strand. 

```{r}
PEReads <- file.path(demo.dir, "demo.sorted.bed.gz")
head(read.table(PEReads, header=FALSE), 2)
```

The output is a data frame in BED file format and/or written to an output BED file, where each line contains the information of a cfDNA fragment.

```{r}
fragInfo <- CollapsePEReads(PEReads)
head(fragInfo, 2)
```

## Collapse methylation states of CpGs on two strands to fragment-level

Function `CollapseCpGs()` generates fragment-level methylation states of CpGs. The main inputs of it are two output files of *bismark methylation extractor*, which is a program performing methylation calling on bisulfite treated sequencing reads. The `CpG_OT*` file contains methylation information for CpGs on the original top strand (OT); the `CpG_OB*` file contains methylation information for CpGs on the original bottom strand (OB). Both files contain columns of sequence ID, methylation state, chromosome name, chromosome start (=chromosome end), methylation call.

```{r}
CpG_OT <- file.path(demo.dir, "CpG_OT_demo.txt.gz")
CpG_OB <- file.path(demo.dir, "CpG_OB_demo.txt.gz")
head(read.table(CpG_OT, header=FALSE), 2)
head(read.table(CpG_OB, header=FALSE), 2)
```
The output is a data frame in BED file format and/or written to an output BED file, where each line contains methylation states of all CpGs on the same fragment. Column *methState* is a sequence of binary values indicating the methylation states of all CpGs on the same fragment (0 represents unmethylated CpG and 1 represents methylated CpG).

```{r}
methInfo <- CollapseCpGs(CpG_OT, CpG_OB)
head(methInfo, 2)
```

## Generate fragment-level information about methylation states

Function `GenerateFragMeth()` combines the output data frame of `CollapsePEReads()` and `CollapseCpGs()` into one data frame, which contains both the fragment information and the methylation states of all CpGs on each fragment. 

```{r}
fragMeth <- GenerateFragMeth(fragInfo, methInfo)
head(fragMeth, 2)
```

## Generate the methylation pattern of markers

Function `GenerateMarkerParam()` calculates paired shape parameters of beta distributions for each marker. There are three main inputs to this function: (1) a data frame of methylation levels (e.g., beta values), where each row is a sample and each column is a marker; (2) a vector of sample types (e.g., tumor or normal, tissue types) corresponding to the rows of the data frame; (3) a vector of marker names corresponding to the columns of the data frame.

```{r}
methLevel <- read.csv(file.path(demo.dir, "beta_matrix.csv"), row.names=1)
sampleTypes <- read.csv(file.path(demo.dir, "sample_type.csv"), row.names=1)$Sample.Type
markerNames <- read.csv(file.path(demo.dir, "marker_index.csv"), row.names=1)$Marker.Index
print(methLevel)
print(sampleTypes)
print(markerNames)
```

The output is a data frame containing the paired shape parameters of beta distributions for markers, which are delimited by `:`. Users can save this data frame into a file with column names, no row names, and columns are delimited by TAB for later use.

```{r}
markerParam <- GenerateMarkerParam(methLevel, sampleTypes, markerNames)
print(markerParam)
```
# Fragments intersecting with marker regions

To make the computation more efficient, users may first only keep the fragments that overlap with the genomic regions of markers. Here, we provide an example of using `R` package `r Biocpkg("GenomicRanges")` to perform the intersection. 

First, transforming the two BED files into GRanges classes.

```{r message=FALSE}
library(GenomicRanges)
demo.dir <- system.file("extdata", package="cfTools")

# a BED file of fragment-level methylation information
frag_bed <- read.csv(file.path(demo.dir, "demo.fragment_level.meth.bed"), 
                     header=TRUE, sep="\t")
frag_meth.gr <- GRanges(seqnames=frag_bed$chr, 
                     ranges=IRanges(frag_bed$start, frag_bed$end),
                     strand=frag_bed$strand,
                     methState=as.character(frag_bed$methState))

# a BED file of genomic regions of markers
markers_bed <- read.csv(file.path(demo.dir, "markers.bed"), 
                        header=TRUE, sep="\t")
markers.gr <- GRanges(seqnames=markers_bed$chr, 
                      ranges=IRanges(markers_bed$start, markers_bed$end),
                      markerName=markers_bed$markerName)

head(frag_meth.gr, 2)
head(markers.gr, 2)
```

Then, overlap two GRanges classes and get the fragment-level methylation states intersecting with the markers.

```{r}
ranges <- subsetByOverlaps(frag_meth.gr, markers.gr, ignore.strand=TRUE)
hits <- findOverlaps(frag_meth.gr, markers.gr,ignore.strand=TRUE)
idx <- subjectHits(hits)

values <- DataFrame(markerName=markers.gr$markerName[idx])
mcols(ranges) <- c(mcols(ranges), values)

marker.methState <- as.data.frame(cbind(ranges$markerName, ranges$methState))
colnames(marker.methState) <- c("markerName", "methState")
head(marker.methState, 4)
```

Alternatively, users may use other softwares like *bedtools intersect* to intersect the BED file of fragment-level information about methylation states generated by `GenerateFragMeth()` with a BED file of genomic regions of markers.

# Cancer detection 

Function `CancerDetector()` separates cfDNA into tumor-derived fragments and background normal fragments and estimates the tumor burden. The main inputs are two files: (1) the fragment-level methylation states of reads (column `methState`) that mapped to the cancer-specific markers; (2) paired shape parameters of beta distributions for cancer-specific markers. All columns are delimited by TAB, and the first line is the column names.

```{r}
fragMethFile <- file.path(demo.dir, "CancerDetector.reads.txt")
markerParamFile <- file.path(demo.dir, "CancerDetector.markers.txt")
head(read.csv(fragMethFile, sep = "\t", colClasses = "character"), 4)
head(read.csv(markerParamFile, sep = "\t"), 4)
```

The output is the estimated tumor burden $\theta$ and the normal cfDNA fraction $1-\theta$.

```{r}
CancerDetector(fragMethFile, markerParamFile)
```

# Tissue deconvolution

Function `cfDeconvolve()` estimates fractions of cfDNA fragments from different tissues (> 2 classes). The main two input files are similar to function `CancerDetector()`: (1) the fragment-level methylation states of reads (column `methState`) that mapped to the tissue-specific markers; (2) paired shape parameters of beta distributions for tissue-specific markers. All columns are delimited by TAB, and the first line is the column names.

```{r}
fragMethFile2 <- file.path(demo.dir, "cfDeconvolve.reads.txt")
markerParamFile2 <- file.path(demo.dir, "cfDeconvolve.markers.txt")
head(read.csv(fragMethFile2, header=TRUE, sep="\t", colClasses = "character"), 4)
head(read.csv(markerParamFile2, header=TRUE, sep="\t", colClasses = "character"), 4)
```
Other input parameters are:

- **Number of tissue types**: a positive integer number *k*. Reads will be classified into *k* different tissue types.
- **Read-based tissue deconvolution EM algorithm type**: options are `em.global.unknown` (default), `em.global.known`, `em.local.unknown`, `em.local.known`.
- **Likelihood ratio threshold**: a positive float number. We suggest this number is 2 (default). All reads with likelihood ratio < cutoff (default: -1.0, meanin no reads filtering is used) will not be used. Likelihood ratio is the max(all tissues' likelihoods)/min(all tissues' likelihoods).
- **EM algorithm maximum iteration number**: a positive integer number. Default is 100.

For example:
```{r}
numTissues <- 7
emAlgorithmType <- "em.global.unknown"
likelihoodRatioThreshold <- 2
emMaxIterations <- 100
```

The output is a data frame containing the cfDNA fractions of different tissue types and an unknown class.

```{r, message=TRUE}
tissueFraction <- cfDeconvolve(fragMethFile2, markerParamFile2, numTissues, emAlgorithmType, likelihoodRatioThreshold, emMaxIterations)
tissueFraction
```


# References
[1] Li W, Li Q, Kang S, Same M, Zhou Y, Sun C, Liu CC, Matsuoka L, Sher L, Wong WH, Alber F, Zhou XJ. CancerDetector: ultrasensitive and non-invasive cancer detection at the resolution of individual reads using cell-free DNA methylation sequencing data. *Nucleic Acids Res*. 2018 Sep 6;46(15):e89. doi: 10.1093/nar/gky423. PMID: 29897492; PMCID: PMC6125664.

[2] Del Vecchio G, Li Q, Li W, Thamotharan S, Tosevska A, Morselli M, Sung K, Janzen C, Zhou X, Pellegrini M, Devaskar SU. Cell-free DNA methylation and transcriptomic signature prediction of pregnancies with adverse outcomes. *Epigenetics*. 2021 Jun;16(6):642-661. doi: 10.1080/15592294.2020.1816774. PMID: 33045922; PMCID: PMC8143248.

# Session info {.unnumbered}
```{r sessionInfo, echo=FALSE}
sessionInfo()
```
