---
title: "Introduction to cfTools"
author: 
  - name: Ran Hu
    affiliation:
      - Department of Pathology and Laboratory Medicine, David Geffen School of Medicine, University of California at Los Angeles
      - Institute for Quantitative & Computational Biosciences, University of California at Los Angeles
      - Bioinformatics Interdepartmental Graduate Program, University of California at Los Angeles
    email: huran@ucla.edu
  - name: Wenyuan Li
    affiliation:
      - Department of Pathology and Laboratory Medicine, David Geffen School of Medicine, University of California at Los Angeles
      - Institute for Quantitative & Computational Biosciences, University of California at Los Angeles
  - name: Xianghong Jasmine Zhou
    affiliation:
      - Department of Pathology and Laboratory Medicine, David Geffen School of Medicine, University of California at Los Angeles
      - Institute for Quantitative & Computational Biosciences, University of California at Los Angeles
package: cfTools
output: 
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
abstract: >
  Cell-free DNA (cfDNA) in blood has emerged as an ideal surrogate for tumor biopsy. It can be obtained noninvasively, and provides a comprehensive landscape of the heterogeneous genetic and epigenetic alterations in tumors. The major bottleneck of cfDNA application is bioinformatics analysis, because cfDNA possesses many unique properties that are not handled properly by general methods. Here, we present a versatile informatics toolset for cfDNA data analysis to facilitate cfDNA-based cancer studies, including (1) Epigenetic mapping: sensitive detection of tumor-derived cfDNA from blood; (2) Measuring tumor-derived cfDNA proportion and inferring the organ of origin noninvasive time-series monitoring.
vignette: >
  %\VignetteIndexEntry{cfTools-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Given the cfMethyl-Seq data of a patient's cell-free DNA (cfDNA) sample, for each cancer marker (or tissue marker), we deconvolve the tumor-derived (or tissue-specific) reads from all reads falling in the marker region. The read-based deconvolution algorithm exploits the pervasiveness of DNA methylation for signal enhancement. Specifically, we generalized upon our previous probabilistic read deconvolution algorithm, i.e., CancerDetector [1], by expanding the 2-class likelihood model to a $T$-class ($T\geq2$) model, for classifying reads into $T$ different classes, and for a given set of markers, we construct a profile vector where the length of the vector is the number of markers and the value in each entry is the normalized counts of tumor-derived (or tissue-derived) reads. This code can deconvolute the different sources of cfDNA reads in two contexts: (1) separating reads into tumor-derived reads and background reads; and (2) separating the reads from different tissues. This algorithm has been applied to two studies for cancer detection [1] and tissue mapping [2].

## Generalized cfDNA sequencing read deconvolution algorithm for both cancer detection and tissue mapping

We aimed to infer $T$-tissue-type composition of a plasma cfDNA sample, which is denoted as a composition vector $\Theta=(\theta_1,\theta_2,\cdots,\theta_T)$ satisfying $0\leq \theta_i < 1$ and $\sum_{t=1}^{T}{\theta_t}=1$ and $\theta_t$ is the cfDNA fraction of tissue t in the plasma. Given a set of $T$-tissue-type methylation signatures denoted as $\Omega$, we used a collection of $N$ cfDNA sequencing reads that were mapped into the genomic regions of all these methylation signatures. This methylation data at individual read level could be represented by a set of $N$ sequencing reads or fragments $\mathbf{R}=({\sf read}_1,{\sf read}_2,\cdots,{\sf read}_N )$, where each read ${\sf read}_i$ was a sequence of binary values (0 or 1) indicating the methylation state of each CpG site covered by the $i$-th read (or fragment). Based on these denotations, we formulated the tissue composition inference problem as a maximum likelihood problem, which estimates the tissue composition $\Theta$ in plasma cfDNAs by maximizing the likelihood $P(\mathbf{R} | \Theta, \Omega)$ or denoted as $P_{\Omega}(\mathbf{R} | \Theta)$. This is formally expressed as shown below:

$$
\begin{align}
\max_{\Theta}\log P_{\Omega}(\mathbf{R} | \Theta) \text{,          s.t.} \sum_{t=1}^{T}{\theta_t}=1 
\end{align}
$$

\noindent where the log-likelihood $\log P_{\Omega}(\mathbf{R} | \Theta)$ is the summation of the log-likelihood of each cfDNA sequencing read:

$$
\begin{align*}
\max_{\Theta}\log P_{\Omega}(\mathbf{R} | \Theta)=\sum_{i=1}^N\log P_{\Omega}({\sf read}_i | \Theta) 
\end{align*}
$$
Since we have $T$ parameters $\theta_1,\theta_2,\cdots,\theta_T$ in this formulation, we cannot use the grid searching for each parameter as it is overly time consuming. Instead, we could employ the Expectation-Maximization (EM) clustering algorithm to solve this problem, where each tissue type could be regarded as a cluster with unknown prior probability $\theta_t$ and each read regarded as an object to be clustered. The EM clustering algorithm typically introduced a missing value or latent random variable $z_i$ for each ${\sf read}_i$ to indicate which tissue type $t$ this read originated from, i.e., $z_i=t$ and $z_i=1,2,\cdots,T$. This latent variable $z_i$ allows the categorical distribution ${\sf Cat}(T,\theta)$, therefore we have $P(z_i=t│\Theta)=\theta_t$. We rewrote $P_\Omega({\sf read}_i│\Theta)$ in the above equation as:

$$
\begin{align*}
	P_{\Omega} \left({\sf read}_i│\Theta \right) &= \sum_{t=1}^T P_\Omega \left({\sf read}_i=t│\Theta \right) \\
	&= \sum_{t=1}^T P\left(z_i=t|\Theta \right) P_\Omega \left({\sf read}_i│z_i=t \right) \\
  &= \sum_{t=1}^T \theta_t P\left({\sf read}_i │ \Omega^t_{\sf marker({\sf read}_i)}\right)
\end{align*}
$$
where $P\left({\sf read}_i│\Omega^t_{{\sf marker(}{\sf read}_i)}\right)$ is the tissue-of-origin likelihood of the cfDNA ${\sf read}_i$ for tissue $t$, and $\Omega^t_{{\sf marker}({\sf read}_i)}$ is the methylation signature of tissue $t$ in the marker’s region that ${\sf read}_i$ is mapped to. Let $q(z_i=t)$ denote the posterior probability of $z_i=t$, i.e., $q(z_i=t)=P(z_i=t|{\sf read}_i,\Theta,\Omega)$. According to the EM algorithm, we deduced the following alternative steps:

$$
\begin{align*}
  \text{E-step:}\qquad & q(z_i=t)=\frac{\theta_t P\left({\sf read}_i | \Omega^t_{{\sf marker}({\sf read}_i)} \right)} {\sum_{t=1}^T \theta_t P\left({\sf read}_i | \Omega^t_{{\sf marker}({\sf read}_i)} \right)} & \text{, for } t=1,\cdots,T \\
  \text{M-step:}\qquad & \theta_t = \frac{\sum^N_{i=1}q(z_i=t)}{N} & \text{, for } t=1,\cdots,T \\
\end{align*}
$$

According to the EM algorithm, starting with a random initial value $\theta$ and iteratively performing the above two equations, this objective function (i.e. the log-likelihood function) converged to a local maximum of the log-likelihood function. We repeated this EM algorithm with different random initialized values of $\theta$, and chose the solution with the maximum log-likelihood. In the E-step, the tissue-of-origin likelihood of a cfDNA ${\sf read}_i$ for tissue $t$ could be easily calculated as $P({\sf read}_i│\Omega)=\Omega^{m({\sf read}_i)} (1-Ω)^{u({\sf read}_i)}$, where $\Omega$ denotes $\Omega^t_{{\sf marker}({\sf read}_i)}$ and $m({\sf read}_i)$ and $u({\sf read}_i)$ are the numbers of methylated and unmethylated CpGs in ${\sf read}_i$, respectively.


# Installation

`r Biocpkg("cfTools")` is an `R` package available via the [Bioconductor](http://bioconductor.org) repository for packages. You can install the release version by using the following commands in your `R` session:

```{r "install", eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}
BiocManager::install("cfTools")
```

Alternatively, you can install the development version from [GitHub](https://github.com/) :

```{r 'install_dev', eval = FALSE}
BiocManager::install("jasminezhoulab/cfTools")
```

# cfDNA sequencing read deconvolution

Example input files are included within the package:
```{r demo}
library(cfTools)
demo.dir <- system.file("extdata", package="cfTools")
```

## Input 1: the cfDNA methylation sequencing reads file

Each line is a read. All columns are delimited by TAB. There is one header line.
We used three column to represent each read's information as below:

- Column 1: marker_index, an integer number that is a marker index.
- Column 2: number of methylated CpG of the read that is mapped to this marker's genomic region.
- Column 3: number of unmethylated CpG of the read that is mapped to this marker's genomic region.

For example:
```{r}
readsBinningFile <- file.path(demo.dir, "example.reads.txt")
readsBinning <- read.table(readsBinningFile, header=TRUE)
head(readsBinning)
```

## Input 2: the markers file

Each line is a marker including its genomic region and methylation pattern (two shape parameters of the beta distribution, which are delimited by `:`). All columns are delimited by TAB. There is one header line.

- Column 1: marker_index
- Column 2+: paired values (shape parameters of the beta distribution) for this marker (each column is a class). Each pair contains two values, delimited by `:`. If you cannot estimate shape parameters of the beta distribution, you may also use a single value, which is the average methylation level of this marker across all samples in the same tissue type.

For example:
```{r}
tissueMarkersFile <- file.path(demo.dir, "example.markers.txt")
tissueMarkers <- read.table(tissueMarkersFile, header=TRUE)
head(tissueMarkers)
```

## Other inputs: 

- **Output type**: options are `tissueFraction` (default), `tissueFraction+readCountRaw`, `tissueFraction+readCountPerMillion`, `tissueFraction+readCountPerBillion`
- **Output file name**: the output will be written to this file.
- **Number of tissue types**: a positive integer number *k*. Reads will be classified into *k* different tissue types.
- **Read-based tissue deconvolution EM algorithm type**: options are `em.global.unknown` (default), `em.global.known`, `em.local.unknown`, `em.local.known`.
- **Likelihood ratio threshold**: a positive float number. We suggest this number is 2 (default). All reads with likelihood ratio < cutoff (default: -1.0, meanin no reads filtering is used) will not be used. Likelihood ratio is the max(all tissues' likelihoods)/min(all tissues' likelihoods).
- **EM algorithm maximum iteration number**: a positive integer number. Default is 100.


For example:
```{r}
outputType <- "tissueFraction+readCountPerBillion"
outputFile <- file.path(demo.dir, "example.profile")
numTissues <- 7
emAlgorithmType <- "em.global.unknown"
likelihoodRatioThreshold <- 2.0
emMaxIterations <- 100
```

## Output: the normalized tissue-specific read counts in each marker

The first three lines are header lines, indicating the column names and the global fraction of tissue types and other information. Starting from the 4-th line, each line corresponds to a marker, consisting of the tissue-specific read counts of this marker.

- Column 1: marker index
- Column 2+: read count of each tissue type in the marker, and the read count of the unknown class is listed in the last column of each line

```{r, message=TRUE}
cfSort(readsBinningFile, tissueMarkersFile, outputType, outputFile, 
       numTissues, emAlgorithmType, likelihoodRatioThreshold, emMaxIterations)

output <- read.table(outputFile, header=TRUE)
head(output)
```


# References
[1] Li W, Li Q, Kang S, Same M, Zhou Y, Sun C, Liu CC, Matsuoka L, Sher L, Wong WH, Alber F, Zhou XJ. CancerDetector: ultrasensitive and non-invasive cancer detection at the resolution of individual reads using cell-free DNA methylation sequencing data. *Nucleic Acids Res*. 2018 Sep 6;46(15):e89. doi: 10.1093/nar/gky423. PMID: 29897492; PMCID: PMC6125664.

[2] Del Vecchio G, Li Q, Li W, Thamotharan S, Tosevska A, Morselli M, Sung K, Janzen C, Zhou X, Pellegrini M, Devaskar SU. Cell-free DNA methylation and transcriptomic signature prediction of pregnancies with adverse outcomes. *Epigenetics*. 2021 Jun;16(6):642-661. doi: 10.1080/15592294.2020.1816774. PMID: 33045922; PMCID: PMC8143248.

# Session info {.unnumbered}
```{r sessionInfo, echo=FALSE}
sessionInfo()
```
